### PR TITLE
Lars op optimiztion with cudaLaunchCooperativeKernel method

### PR DESCRIPTION
**Pull request by JamesLim-sy**
### PR types

Performance optimization
### PR changes

OPs
### Describe

- Conception 
Theoretically,   the lars op was consist of 2 steps :
    - L2 norm of grad and param respectively to generate fundamental hyper-param for lars update.
    - Lars update.

- Drawback:
Originally, the L2 norm kernel was implemented with combined eigen kernel below. As shown by codes, a combined L2 norm kernel relied on two eigen kernels. Hence, while training with lars optimizer,  eigen kernel would be called 4 times in total. Take  lars update cuda kernel into account,  5 kernel were needed for only 1 lars op.
```
     eigen_p.template cast<MPDType>().square().sum().sqrt();
    (eigen_g.template cast<MPDType>() * rescale_grad).square().sum().sqrt();
```
截屏2021-09-06 下午7 13 34

- Optimization method
 1.  Instead of  4 eigen kernels, whole L2 norm calculation can be combined into only 1 cuda kernel, 
 2. L2 norm is acquired with two steps : 
     - (1)  Partial reduction result of pow(grad, 2) and pow(param, 2) in each block and store into temporary global memory.
     - (2) Read partial reduction results from gloabl memory and reduce to get the final l2 norm result.
 3. No more than 1024 * sizeof(DataType)  temporary  memory was needed for partial reduction in L2 norm by design,  in case of too much temporary memory occupied.
 4. Apart from L2 norm,  lars update part was optimized with vectorize IO to gain the performance.
 5. Once running on **A100 device**, **whole lars op be combined into just 1 cuda kernel**, with the help of `cudaLaunchCooperativeKernel`. Otherwise, two kernels are needed, namely step (1) was estabulished with one kernel, step (2) and lars update were estabulished with another kernel.

- Performance 
1. After optimization of lars op **without** `cudaLaunchCooperativeKernel`, the training performance increase from 2969 ips to 3040 ips, and total kernel needed shrink from 5 to 2.
image

2. After optimization of lars op **with** `cudaLaunchCooperativeKernel`, the training performance increase from 2969 ips to 3110.58 ips, , and total kernel needed shrink from 5 to 1.
截屏2021-09-11 下午5 05 24
 

- Note 
More than `cudaLaunchCooperativeKernel` or `DynamicParallelism`,  five other methods acquired from internet had been tried, but none of them worked well in CUDA kernel combination. So, I introduce these 2 ways mentioned to you, my friends. They truly may help.

- Convergence 
Loss comparison while trainning with or without perf-optimization of lars kernel in one CUDA card are shown in figure below:
截屏2021-09-24 下午3 11 49

As for resnet50, converged at epoch 35, convergence stats data are shown below:
>  loss 29.42091, acc1 0.75918, acc5 0.92844, time 17.7330 sec

### CODE REVIEW

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 1]
```diff
@@ -14,72 +14,276 @@ limitations under the License. */
...
+#ifdef __HIPCC__
+#define LARS_BLOCK_SIZE 256
+#else
+#define LARS_BLOCK_SIZE 512
+#endif
+
+#if CUDA_VERSION >= 11000
+#define FUNCTION_FLAG __device__
+#else
+#define FUNCTION_FLAG __global__
```
Comment (ZzSean) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: global函数内还能再调用global吗

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 2]
```diff
@@ -14,72 +14,276 @@ limitations under the License. */
...
+#endif
 
 namespace paddle {
 namespace operators {
 
 template <typename T>
 using MultiPrecisionType = typename details::MPTypeTrait<T>::Type;
 
+__device__ __forceinline__ float SquareRoot(float x) { return sqrtf(x); }
+__device__ __forceinline__ double SquareRoot(double x) { return sqrt(x); }
```
Comment (ZzSean) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: cuda内的sqrt本身已经重载过float类型了，可以不用再封装一层

Comment (ZzSean) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:44]: cuda内的sqrt本身已经重载过float类型了，可以不用再封装一层

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 3]
```diff
@@ -14,72 +14,276 @@ limitations under the License. */
...
+__device__ __forceinline__ float FmaRoot(float x, float y, float z) {
```
Comment (ZzSean) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 同上

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 4]
```diff
@@ -14,72 +14,276 @@ limitations under the License. */
...
+  return fmaf(x, y, z);
+}
+__device__ __forceinline__ double FmaRoot(double x, double y, double z) {
+  return fma(x, y, z);
+}
+
+template <typename MT, int VesSize>
```
Comment (ZzSean) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: VesSize->VecSize

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 5]
```diff
@@ -14,72 +14,276 @@ limitations under the License. */
...
-    const MultiPrecisionType<T>* learning_rate, const MT mu, const int64_t num,
-    const MT lars_coeff, const MT lars_weight_decay,
-    const MultiPrecisionType<T>* p_norm, const MultiPrecisionType<T>* g_norm,
-    T* p_out, MT* v_out, const MT epsilon, const MT* master_p, MT* master_p_out,
-    const MultiPrecisionType<T> rescale_grad) {
-  const MT lr = static_cast<MT>(learning_rate[0]);
-  MT local_lr = lr;
-  const MT p_n = static_cast<MT>(p_norm[0]);
-  const MT g_n = static_cast<MT>(g_norm[0]);
+__device__ inline void VectorizeLarsUpdateMP(
```
Comment (ZzSean) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: VectorizeLarsUpdateMP和VectorizeLarsUpdate不可以写成一个吗

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 1 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这里主要是考虑到代码的复用性，对于`CUDA >= 1100`即CUDA 11的情况， `L2NormKernel` 可以合并至`MomentumUpdateKernel` 内实现，此时是`__device__ kernel`；其他情况下，与`MomentumUpdateKernel` 分离，成为单独的`__gloabl__ kernel` 。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 2 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 好的，根据建议修改。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 3 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 根据建议修改。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 4 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 根据建议修改。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 5 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 封装后的代码太乱了，所以就拆开分成了两份写。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 6]
```diff
@@ -14,72 +14,262 @@ limitations under the License. */
 
 #include "paddle/fluid/framework/op_registry.h"
 #include "paddle/fluid/operators/amp/fp16_type_traits.h"
+#include "paddle/fluid/operators/math/math_cuda_utils.h"
 #include "paddle/fluid/operators/optimizers/lars_momentum_op.h"
+#include "paddle/fluid/platform/fast_divmod.h"
+
+#if defined(__NVCC__) && CUDA_VERSION >= 11000
+#include <cooperative_groups.h>
+#define LARS_FUNCTION_FLAG __device__
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:30]: 加些注释说明下，为什么这样定义

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 7]
```diff
@@ -14,72 +14,262 @@ limitations under the License. */
...
 
 namespace paddle {
 namespace operators {
 
 template <typename T>
 using MultiPrecisionType = typename details::MPTypeTrait<T>::Type;
 
+template <typename MT, int VecSize>
+__device__ inline void VectorizeLarsUpdate(
+    const MT* __restrict__ g, const MT* __restrict__ v, MT* __restrict__ p_out,
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: p、g、v尽量用全称吧，param、grad、velocity

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 8]
```diff
@@ -14,72 +14,262 @@ limitations under the License. */
...
+
+  for (int i = tid; i < main; i += grid_stride) {
+    VecMType v_new, p_new;
+    VecMType g_data = g_vec[i];
+    VecMType v_data = v_vec[i];
+    VecMType p_data = p_vec[i];
+
+#pragma unroll
+    for (int j = 0; j < VecSize; ++j) {
+      MT grad = g_data.val[j] * rescale_grad;
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 可直接用`g_data[j]`访问

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 9]
```diff
@@ -14,72 +14,262 @@ limitations under the License. */
...
+#define LARS_BLOCK_SIZE 512
+#endif
 
 namespace paddle {
 namespace operators {
 
 template <typename T>
 using MultiPrecisionType = typename details::MPTypeTrait<T>::Type;
 
+template <typename MT, int VecSize>
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 这里用`T`吧

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 5 above)
Comment (Xreki) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 我也觉得封装成一个函数比较好，代码相似度很高，理论上并不会“太乱”。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 10]
```diff
@@ -14,72 +14,262 @@ limitations under the License. */
...
+__device__ inline void VectorizeLarsUpdateMP(
+    const T* __restrict__ g, const MT* __restrict__ v, T* __restrict__ p_out,
+    MT* __restrict__ v_out, const MT* __restrict__ master_p,
+    MT* __restrict__ master_p_out, const MT mu, MT local_lr,
+    const MT lars_weight_decay, const MT rescale_grad, const int tid,
+    const int grid_stride, const int numel) {
+  // As for multiple-precision, type T and MT cannot be more than fp16 or fp32,
+  // Then, the maximum data IO size could be set to 4.
+  using VecType = paddle::platform::AlignedVector<T, 4>;
+  using VecMType = paddle::platform::AlignedVector<MT, 4>;
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 为什么强制写死成了`4`？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 11]
```diff
@@ -14,72 +14,262 @@ limitations under the License. */
...
+        lr * lars_coeff * p_n / (fma(lars_weight_decay, p_n, g_n) + epsilon);
+  }
+
+  if (master_p) {
+    VectorizeLarsUpdateMP<T, MT>(g, v, p_out, v_out, master_p, master_p_out, mu,
+                                 local_lr, lars_weight_decay, rescale_grad, tid,
+                                 grid_stride, numel);
+  } else {
+    if (std::is_same<T, float>::value ||
+        std::is_same<T, paddle::platform::float16>::value) {
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:225]: 不用判断地址是否对齐吗？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 12]
```diff
@@ -105,41 +291,101 @@ class LarsMomentumOpCUDAKernel : public framework::OpKernel<T> {
...
+    int num_blocks_per_sm = 0;
+    cudaOccupancyMaxActiveBlocksPerMultiprocessor(&num_blocks_per_sm,
+                                                  MomentumLarsKernel<T, MT>,
+                                                  LARS_BLOCK_SIZE, sizeof(MT));
+    int sm_num = cuda_ctx.GetSMCount();
+    int grid_real =
+        std::min(std::min(sm_num * num_blocks_per_sm, grid), LARS_BLOCK_SIZE);
+    framework::Tensor tmp_buffer_t =
+        ctx.AllocateTmpTensor<MT, platform::CUDADeviceContext>(
+            {LARS_BLOCK_SIZE << 1}, cuda_ctx);
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:321]: 协作组方式也需要临时空间吗？

**Review by Xreki — COMMENTED** [review]
验证一下ResNet50训练的收敛情况，loss、训练精度变化画成图贴到PR描述中。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 6 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:30]: 好的，根据建议添加注释。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 7 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 根据建议修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 9 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 根据建议修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 8 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 好的，根据建议修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 10 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 对于FP16的数据而言，T类型代表fp16, MT 类型代表 fp32，单次向量化IO 4个数据规模并不会超过128 bits的限制，所以这里显式地写明是4了，后续会添加注释

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 12 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:321]: 申请临时空间用来存储各block上的parital_l2_nrom，为后续计算全局l2_norm做准备，引入协作组的`Cooperative_groups`的目的是达成 grid_sync操作

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 5 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu]: 好的，根据建议修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 13]
```diff
@@ -14,72 +14,248 @@ limitations under the License. */
...
+__device__ __forceinline__ float Sqrt(float x) { return sqrtf(x); }
+__device__ __forceinline__ double Sqrt(double x) { return sqrt(x); }
+__device__ __forceinline__ float Fma(float x, float y, float z) {
+  return fmaf(x, y, z);
+}
+__device__ __forceinline__ double Fma(double x, double y, double z) {
+  return fma(x, y, z);
+}
+
+template <typename T, typename MT, int VecSize, bool IsAmp = false>
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:56]: 没有必要增加`IsAmp`这样一个模板变量，可以判断`T`是不是`float16`类型。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 14]
```diff
@@ -14,72 +14,248 @@ limitations under the License. */
...
+__device__ inline void VectorizeLarsUpdate(
+    const T* __restrict__ grad, const MT* __restrict__ param,
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:58]: `param`其实是`master_param`？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 15]
```diff
@@ -14,72 +14,248 @@ limitations under the License. */
...
+    VecMType velocity_data = velocity_vec[i];
+
+#pragma unroll
+    for (int j = 0; j < VecSize; ++j) {
+      MT grad_val = static_cast<MT>(grad_data[j]) * rescale_grad;
+      velocity_tmp[j] =
+          Fma(velocity_data[j], mu,
+              local_lr * Fma(lars_weight_decay, param_data[j], grad_val));
+      param_tmp[j] = param_data[j] - velocity_tmp[j];
+      param_out_tmp[j] = static_cast<T>(param_tmp[j]);
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:96]: 这两个变量的命名不太直观。`param_tmp`应该是`master_param_out_tmp`。

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu [hunk 16]
```diff
@@ -14,72 +14,248 @@ limitations under the License. */
...
+      reinterpret_cast<const VecMType*>(param);
+  const VecMType* __restrict__ velocity_vec =
+      reinterpret_cast<const VecMType*>(velocity);
+  VecType* param_out_vec = reinterpret_cast<VecType*>(param_out);
+  VecMType* velocity_out_vec = reinterpret_cast<VecMType*>(velocity_out);
+
+  VecMType* master_param_out_vec;
+  if (IsAmp) {
+    master_param_out_vec = reinterpret_cast<VecMType*>(master_param_out);
+  }
```
Comment (Xreki) [inline paddle/fluid/operators/optimizers/lars_momentum_op.cu:80]: `_vec`我会以为是`AlignedVector`类型，结果却是指针。用aligned_vector.h里面封装的`Load`、`Store`函数，会不会好一些？

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 11 above)
Comment (Xreki) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:225]: 这个分支还可能是`float16`类型？

**Review by Xreki — DISMISSED** [review]
LGTM. 如果精度验证没有问题，这个PR可以先合进去。

**Review by Xreki — APPROVED** [review]

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 15 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:96]: 下个commit会根据建议修改

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 14 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:58]: 对于AMP计算来说，传入的是`master_param`；对于非AMP计算，传入的直接就是`param`了，这里的形参统一用了`param`命名

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 16 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:80]: 命名方式确实存在问题

File: paddle/fluid/operators/optimizers/lars_momentum_op.cu (hunk 13 above)
Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:56]: 之前修改这里比较疑惑的问题是：调用`VectorizeLarsUpdate`接口的时候，会不会存在针对fp16类型数据也存在`IsAmp=false`的情形。所以，当时没有用std::is_same<> 转用了加入一个模板参数。

Comment (JamesLim-sy) [reply paddle/fluid/operators/optimizers/lars_momentum_op.cu:56]: 又看了一次lars_oprimizer 的构造过程，其中包含一条：
https://github.com/PaddlePaddle/Paddle/blob/1f93582cd1f13a09971e2c03334d649d82238e5b/python/paddle/fluid/optimizer.py#L2061-L2093
看逻辑是必须同时满足`type == fp16，IsAmp == true` 的情况下才能用`master_param`，单独用`float16`类型作判断应该是不够，但是不知道这里是不是amp部分设计的逻辑有问题，我去问问amp的同学。

**Pull request merged**